File: cbam/cbam/page/financial_evaluation_report/financial_evaluation_report.py
import frappe
import json

@frappe.whitelist()
def get_report_data(filters=None, selected_filters=None, start=0, page_length=50):
    if filters and isinstance(filters, str):
        filters = json.loads(filters)
    filters = filters or {}
    if selected_filters and isinstance(selected_filters, str):
        selected_filters = json.loads(selected_filters)
    selected_filters = selected_filters or {}
    try:
        start = int(start)
    except Exception:
        start = 0
    try:
        page_length = int(page_length)
    except Exception:
        page_length = 50

    columns = get_columns()

    data, total_count = get_data(filters, selected_filters, start, page_length)
    chart_data = get_chart_data(data)

    return {"columns": columns, "data": data, "chart_data": chart_data, "total_count": total_count}


def get_columns():
    return [
        {"fieldname": "cn_code", "fieldtype": "Data", "label": "CN Code", "width": 150},
        {"fieldname": "article_number", "fieldtype": "Data", "label": "Article Number", "width": 200},
        {"fieldname": "supplier", "fieldtype": "Data", "label": "Supplier", "width": 200},
        {"fieldname": "raw_mass", "fieldtype": "Data", "label": "Mass [kg]", "width": 150},
        {"fieldname": "carbon_price_due", "fieldtype": "Data", "label": "Carbon Price Due", "width": 150},
        {"fieldname": "installation_country", "fieldtype": "Data", "label": "Installation Country", "width": 180},
        {"fieldname": "real_emission_value", "fieldtype": "Data", "label": "Specific (Direct) Emission Value", "width": 250},
        {"fieldname": "real_emission_cost", "fieldtype": "Data", "label": "Actual Cost", "width": 180},
        {"fieldname": "standard_emission_cost", "fieldtype": "Data", "label": "Standard Cost", "width": 180},
    ]


def get_data(filters=None, selected_filters=None, start=0, page_length=50):
    filters = filters or {}
    selected_filters = selected_filters or {}
    try:
        start = int(start)
    except Exception:
        start = 0
    try:
        page_length = int(page_length)
    except Exception:
        page_length = 50

    # Merge selected_filters into filters, but only non-empty values
    for key, value in selected_filters.items():
        if value is not None and value != '' and value != []:
            if isinstance(value, list):
                if len(value) > 0:
                    filters[key] = value
            else:
                filters[key] = value

    where_clauses = []
    where_clauses_eg = []

    user = frappe.session.user
    declarants = get_declarant_for_user(user)
    if declarants:
        declarant_list = ', '.join(f"'{d}'" for d in declarants)
        where_clauses.append(f"g.declarant IN ({declarant_list})")

    if filters.get("cn_code") and isinstance(filters["cn_code"], list) and len(filters["cn_code"]) > 0:
        cn_code_list = ', '.join(f"'{c}'" for c in filters["cn_code"])
        where_clauses.append(f"(g.cn_code IN ({cn_code_list}))")
        where_clauses_eg.append(f"(eg.cn_code IN ({cn_code_list}))")

    if filters.get("supplier") and isinstance(filters["supplier"], list) and len(filters["supplier"]) > 0:
        supplier_list = ', '.join(f"'{s}'" for s in filters["supplier"])
        where_clauses.append(f"(g.supplier_name IN ({supplier_list}))")
        where_clauses_eg.append(f"(eg.supplier IN ({supplier_list}))")

    if filters.get("article_number") and isinstance(filters["article_number"], list) and len(filters["article_number"]) > 0:
        article_number_list = ', '.join(f"'{s}'" for s in filters["article_number"])
        where_clauses.append(f"(g.article_number IN ({article_number_list}))")
        where_clauses_eg.append(f"(eg.article_no IN ({article_number_list}))")
    
    if filters.get("reporting_period") and isinstance(filters["reporting_period"], list) and len(filters["reporting_period"]) > 0:
        reporting_period_list = ', '.join(f"'{s}'" for s in filters["reporting_period"])
        where_clauses.append(f"(g.internal_customs_import_number IN ({reporting_period_list}))")
        where_clauses_eg.append(f"(eg.reporting_period IN ({reporting_period_list}))")

    where_sql = ""
    if where_clauses:
        where_sql = "WHERE " + " AND ".join(where_clauses)
        
    where_sql_eg = ""
    if where_clauses_eg:
        where_sql_eg = "WHERE " + " AND ".join(where_clauses_eg) if where_clauses_eg else ""

    cbam_factor = float(selected_filters.get('cbam_factor', 0.0) or 0.0)
    bench_mark = float(selected_filters.get('bench_mark', 0.0) or 0.0)
    emission_value = float(selected_filters.get('emission_value', 0.0) or 0.0)
    ets_carbon_price = float(selected_filters.get('ets_price_value', 0.0) or selected_filters.get('ets_price', 0.0) or 0.0)

    # Count total rows for pagination
    count_query = f"""
        SELECT COUNT(*) FROM (
            SELECT 
                eg.cn_code, eg.article_no as article_number, eg.supplier, eg.raw_mass, eg.installation_country,
                eg.specific_direct_embedded_emissions, eg.carbon_price_due
            FROM `tabExternal Good` eg
            {where_sql_eg}
            
            UNION
            SELECT 
                g.cn_code, g.article_number, g.supplier_name as supplier, g.raw_mass, g.installation_country,
                g.specific_direct_embedded_emissions, g.carbon_price_due
            FROM `tabGood` g
            {where_sql}
        ) AS count_table

    """
    total_count = frappe.db.sql(count_query)[0][0]

    # Main data query with LIMIT/OFFSET for pagination
    data_query = f"""
        SELECT * FROM (
            SELECT 
                eg.cn_code,
                eg.article_no AS article_number,
                eg.supplier,
                IFNULL(eg.raw_mass,0.0) AS raw_mass,
                IFNULL(eg.carbon_price_due,0.0) as carbon_price_due,
                eg.installation_country,
                IFNULL(eg.specific_direct_embedded_emissions,0.0) AS real_emission_value,             
                ((
                    IFNULL(eg.specific_direct_embedded_emissions,0.0) - ({cbam_factor} * {bench_mark})
                    - ((IFNULL(eg.specific_direct_embedded_emissions,0.0) * IFNULL(eg.carbon_price_due, 0.0)) / {ets_carbon_price})
                ) * IFNULL(eg.raw_mass, 0.0) * {ets_carbon_price}) AS real_emission_cost,

                ((
                    {emission_value}
                    - ({bench_mark} * {cbam_factor})
                    - (({emission_value} * IFNULL(eg.carbon_price_due, 0.0)) / {ets_carbon_price})
                ) * IFNULL(eg.raw_mass, 0.0) * {ets_carbon_price}) AS standard_emission_cost
            FROM `tabExternal Good` eg
            {where_sql_eg}
            UNION
            SELECT 
                g.cn_code,
                g.article_number,
                g.supplier_name AS supplier,
                IFNULL(g.raw_mass,0.0) AS raw_mass,
                IFNULL(g.carbon_price_due,0.0) AS carbon_price_due,
                g.installation_country,    
                IFNULL(g.specific_direct_embedded_emissions,0.0) AS real_emission_value,
                ((
                    IFNULL(g.specific_direct_embedded_emissions,0.0) - ({cbam_factor} * {bench_mark})
                    - ((IFNULL(g.specific_direct_embedded_emissions,0.0) * IFNULL(g.carbon_price_due,0.0)) / {ets_carbon_price})
                ) * IFNULL(g.raw_mass,0.0) * {ets_carbon_price}) AS real_emission_cost,
                ((
                    {emission_value}
                    - ({bench_mark} * {cbam_factor})
                    - (({emission_value} * IFNULL(g.carbon_price_due, 0.0)) / {ets_carbon_price})
                )
                * IFNULL(g.raw_mass, 0.0) * {ets_carbon_price}) AS standard_emission_cost
            FROM `tabGood` g
            {where_sql}
        ) AS main_table
        LIMIT {page_length} OFFSET {start}
    """
    data = frappe.db.sql(data_query, as_dict=1)
    frappe.log_error("121test",{
        "data_length": len(data),
        "total_count": total_count,
        "start": start,
        "page_length": page_length
    })
    return data, total_count


def get_chart_data(data):
    """Process data to create chart data grouped by year"""
    year_data = {}
    
    for row in data:
        # Extract year from reporting_period or use current year as fallback
        year = None
        if row.get('reporting_period'):
            try:
                # Try to extract year from reporting_period field
                year = str(row['reporting_period'])[:4]  # Take first 4 characters as year
            except:
                pass
        
        if not year:
            # Fallback to current year if no reporting_period
            from datetime import datetime
            year = str(datetime.now().year)
        
        if year not in year_data:
            year_data[year] = {
                'actual_cost': 0.0,
                'standard_cost': 0.0,
                'count': 0
            }
        
        # Sum up the costs for each year
        year_data[year]['actual_cost'] += float(row.get('real_emission_cost', 0.0) or 0.0)
        year_data[year]['standard_cost'] += float(row.get('standard_emission_cost', 0.0) or 0.0)
        year_data[year]['count'] += 1
    
    # Convert to chart format
    chart_data = {
        'labels': [],
        'actual_costs': [],
        'standard_costs': []
    }
    
    # Sort by year
    for year in sorted(year_data.keys()):
        chart_data['labels'].append(year)
        chart_data['actual_costs'].append(year_data[year]['actual_cost'])
        chart_data['standard_costs'].append(year_data[year]['standard_cost'])
    
    return chart_data


def get_declarant_for_user(user):
    result = frappe.db.get_all(
        "Declarant User",
        filters={"user": user},
        fields=["parent"],
        distinct=True
    )
    return [row.parent for row in result]  # List of declarant names

@frappe.whitelist()
def get_cards_value(filters=None):
    filters = json.loads(filters)

    filters = filters or {}

    sql = """
        SELECT 
            cf.cbam_factor,
            cnb.bench_mark,
            sev.emission_value,
            ecp.price AS ets_price
        FROM `tabCBAM Factor` cf
        LEFT JOIN `tabCN Code Bench Mark Emission Value` cnb 
            ON cnb.year = %(year)s
        LEFT JOIN `tabStandard Emission Value` sev 
            ON sev.year = %(year)s
        LEFT JOIN `tabETS Carbon Price` ecp 
            ON YEAR(ecp.price_date) = %(year)s AND ecp.ets_price_type = %(ets_price_type)s
        WHERE cf.year = %(year)s
        LIMIT 1
    """

    result = frappe.db.sql(sql, filters, as_dict=True)
    return result[0] if result else {}

File: cbam/cbam/page/financial_evaluation_report/financial_evaluation_report.js
// Copyright (c) 2024, CBAM Team. All rights reserved.
// MIT License. See license.txt

/**
 * Financial Evaluation Report Page
 * Optimized and refactored for Frappe best practices and maintainability.
 */

frappe.provide('cbam.pages');

frappe.pages['financial-evaluation-report'].on_page_load = function (wrapper) {
    (function () {
        const page = frappe.ui.make_app_page({
            parent: wrapper,
            title: __('Financial Evaluation Report'),
            single_column: true,
        });

        // Layout
        render_layout(page.body);
        cbam.inject_custom_styles(page.body);

        // Filters
        const filters = setup_filters();
        cbam.setup_filter_clear_buttons(filters, load_report_table);
        cbam.setup_filter_listeners(filters, load_report_table, update_stat_card_values, (ets_price_type, year) => cbam.refresh_ets_price_options(filters, ets_price_type, year));

        // Pagination state
        let start = 0;
        let page_length = 50;
        let total_count = 0;
        let all_data = [];
        let datatable = null;

        // Initial Stats, Chart, and Table
        clear_stats();
        cbam.update_chart(null);
        load_report_table(true);

        // Setup table toggle functionality
        setup_table_toggle();

        // Helper: Get table data, optionally per tonne
        function get_table_data(data, per_tonne = false) {
            return data.map(row => {
                if (!per_tonne) return { ...row };
                let mass_tonnes = Number(row.raw_mass_tonne) || 0;
                if (mass_tonnes > 0) {
                    return {
                        ...row,
                        real_emission_cost: (Number(row.real_emission_cost) || 0) / mass_tonnes,
                        standard_emission_cost: (Number(row.standard_emission_cost) || 0) / mass_tonnes
                    };
                } else {
                    return {
                        ...row,
                        real_emission_cost: 0,
                        standard_emission_cost: 0
                    };
                }
            });
        }

        // --- Show Selected Rows Toggle Logic ---
        function get_selected_rows_data() {
            if (!datatable) return [];
            const selectedIndexes = datatable.rowmanager.getCheckedRows();
            return selectedIndexes.map(idx => all_data[idx]);
        }

        // Listen for 'Show Cost per tonne' toggle event
        $('#table-toggle').on('change', function() {
            const per_tonne = $(this).is(':checked');
            // Update table
            const table_data = get_table_data(all_data, per_tonne);
            if (datatable) {
                datatable.refresh(table_data);
            }
            // Update chart (respect 'Show Selected rows' toggle)
            const showSelected = $('#selected-rows-toggle').is(':checked');
            let chart_data;
            if (showSelected) {
                const selectedData = get_selected_rows_data();
                chart_data = cbam.get_chart_data(selectedData, per_tonne);
            } else {
                chart_data = cbam.get_chart_data(all_data, per_tonne);
            }
            cbam.update_chart(chart_data);
        });

        // Listen for 'Show Selected rows' toggle event
        $('#selected-rows-toggle').on('change', function() {
            const showSelected = $(this).is(':checked');
            const per_tonne = $('#table-toggle').is(':checked');
            let chart_data;
            if (showSelected) {
                const selectedData = get_selected_rows_data();
                chart_data = cbam.get_chart_data(selectedData, per_tonne);
            } else {
                chart_data = cbam.get_chart_data(all_data, per_tonne);
            }
            cbam.update_chart(chart_data);
        });

        // Update chart live when selection changes if 'Show Selected rows' is ON
        function bind_datatable_selection_events() {
            if (!datatable) return;
            datatable.on('onCheckRow', function() {
                if ($('#selected-rows-toggle').is(':checked')) {
                    const selectedData = get_selected_rows_data();
                    const per_tonne = $('#table-toggle').is(':checked');
                    cbam.update_chart(cbam.get_chart_data(selectedData, per_tonne));
                }
            });
            datatable.on('onUncheckRow', function() {
                if ($('#selected-rows-toggle').is(':checked')) {
                    const selectedData = get_selected_rows_data();
                    const per_tonne = $('#table-toggle').is(':checked');
                    cbam.update_chart(cbam.get_chart_data(selectedData, per_tonne));
                }
            });
        }

        /**
         * Render the main page layout.
         * @param {HTMLElement} body
         */
        function render_layout(body) {
            $(body).html(`
                <div class="container p-0">
                    <div class="row mb-3 align-items-center">
                        <div class="col">
                            <div class="section-title mb-2">🧾 Select Article</div>
                            <div class="row gx-2" id="filter-section-group-1"></div>
                        </div>
                        <div class="col-auto d-flex align-items-center" style="margin-top: 22px;">
                            <button class="btn btn-primary btn-xs clear-selections text-nowrap filter-clear-btn" id="clear-group-1">Clear All</button>
                        </div>
                    </div>
                    <div class="row mb-4 align-items-center">
                        <div class="col">
                            <div class="section-title mb-2">📅 Calculation Based On</div>
                            <div class="row gx-2" id="filter-section-group-2"></div>
                        </div>
                        <div class="col-auto d-flex align-items-center" style="margin-top: 22px;">
                            <button class="btn btn-primary btn-xs clear-selections text-nowrap filter-clear-btn" id="clear-group-2">Clear All</button>
                        </div>
                    </div>
                    <div class="row mb-4" id="card-section">
                        <div class="row gx-3" id="compact-stat-row"></div>
                    </div>
                    <div class="frappe-card mb-4" id="chart-section"></div>
                    <div class="frappe-card mb-4" id="table-toggle-section">
                        <div class="p-2 d-flex align-items-center gap-5">
                            <span class="ms-2 small">Show Cost per tonne</span>&nbsp;
                            <label class="switch-compact mb-0 me-4 ms-2">
                                <input type="checkbox" id="table-toggle">
                                <span class="slider-compact round"></span>
                            </label>
                            <span class="vr mx-3"></span>
                            <span class="ms-2 small">Show Selected rows</span>&nbsp;
                            <label class="switch-compact mb-0 ms-2">
                                <input type="checkbox" id="selected-rows-toggle">
                                <span class="slider-compact round"></span>
                            </label>
                        </div>
                    </div>
                    <div class="frappe-card mb-4" id="table-scroll-container" style="overflow-x: auto;">
                        <div id="table-section"></div>
                    </div>
                </div>
            `);
        }

        /**
         * Setup all filters and return the filter controls object.
         * @returns {Object}
         */
        function setup_filters() {
            const currentYear = frappe.datetime.get_today().split('-')[0];
            const filters = {
                cn_code: cbam.create_filter('CN Code', 'MultiSelectList', 'cn_code', '#filter-section-group-1', []),
                supplier: cbam.create_filter('Supplier', 'MultiSelectList', 'supplier', '#filter-section-group-1', []),
                article_number: cbam.create_filter('Article Number', 'MultiSelectList', 'article_number', '#filter-section-group-1', []),
                reporting_period: cbam.create_filter('Reporting Period', 'MultiSelectList', 'reporting_period', '#filter-section-group-1', []),
                year: cbam.create_filter('Year', 'Link', 'year', '#filter-section-group-2', null, 'Year', currentYear),
                ets_price_type: cbam.create_filter('ETS Price Type', 'Select', 'ets_price_type', '#filter-section-group-2', ['', 'Actual', 'Prediction'], null, 'Actual'),
                ets_price: cbam.create_filter('ETS Price', 'Select', 'ets_price', '#filter-section-group-2', []),
            };

            // CN Code: no dependencies
            filters.cn_code.df.get_data = function(txt) {
                return frappe.db.get_list('CN Code', {
                    fields: ['cn_code as value', 'cn_code as description'],
                    filters: [['cn_code', 'like', `%${txt}%`]],
                    limit: 20,
                });
            };

            // Supplier: depends on selected CN Code(s)
            filters.supplier.df.get_data = function(txt) {
                const cn_code_vals = filters.cn_code?.get_value?.() || [];
                let filterArr = [['supplier', 'like', `%${txt}%`]];
                if (cn_code_vals.length) {
                    filterArr.push(['cn_code', 'in', cn_code_vals]);
                }
                return frappe.db.get_list('External Good', {
                    fields: ['supplier as value', 'supplier as description'],
                    filters: filterArr,
                    distinct: true,
                    limit: 20,
                });
            };

            // Article Number: depends on selected Supplier(s) and CN Code(s)
            filters.article_number.df.get_data = function(txt) {
                const supplier_vals = filters.supplier?.get_value?.() || [];
                const cn_code_vals = filters.cn_code?.get_value?.() || [];
                let filterArr = [['article_no', 'like', `%${txt}%`]];
                if (supplier_vals.length) {
                    filterArr.push(['supplier', 'in', supplier_vals]);
                }
                if (cn_code_vals.length) {
                    filterArr.push(['cn_code', 'in', cn_code_vals]);
                }
                return frappe.db.get_list('External Good', {
                    fields: ['article_no as value', 'article_no as description'],
                    filters: filterArr,
                    distinct: true,
                    limit: 20,
                });
            };

            // Reporting Period: depends on selected Supplier(s) and CN Code(s)
            filters.reporting_period.df.get_data = function(txt) {
                const supplier_vals = filters.supplier?.get_value?.() || [];
                const cn_code_vals = filters.cn_code?.get_value?.() || [];
                let filterArr = [['reporting_period', 'like', `%${txt}%`]];
                if (supplier_vals.length) {
                    filterArr.push(['supplier', 'in', supplier_vals]);
                }
                if (cn_code_vals.length) {
                    filterArr.push(['cn_code', 'in', cn_code_vals]);
                }
                return frappe.db.get_list('External Good', {
                    fields: ['reporting_period as value', 'reporting_period as description'],
                    filters: filterArr,
                    distinct: true,
                    limit: 20,
                });
            };

            return filters;
        }

        /**
         * Update stat card values based on filter changes.
         * @param {string} key
         * @param {Object} filters
         */
        function update_stat_card_values(key, filters) {
            if (key === 'ets_price') {
                $('.stat-value[data-stat="ets-price"]').text(filters.ets_price || 0.0);
                return;
            }
            clear_stats();
            frappe.call({
                method: 'cbam.cbam.page.financial_evaluation_report.financial_evaluation_report.get_cards_value',
                args: { filters },
                callback: function (r) {
                    if (!r.message) return;
                    const updated_values = {
                        'standard-emission-value': r.message.emission_value || 0.0,
                        'bench-mark-emission-value': r.message.bench_mark || 0.0,
                        'cbam-factor': r.message.cbam_factor || 0.0,
                        'ets-price': r.message.ets_price || 0.0,
                        'year': filters.year || frappe.datetime.get_today().split('-')[0],
                    };
                    Object.entries(updated_values).forEach(([key, val]) => {
                        $('.stat-value[data-stat="' + key + '"]').text(val);
                    });
                },
                error: function () {
                    clear_stats();
                }
            });
        }

        /**
         * Clear and render stat cards with default values.
         */
        function clear_stats() {
            const compact_stats = [
                { label: 'Standard Emission Value', value: '--' },
                { label: 'Bench Mark Emission Value', value: '--' },
                { label: 'CBAM Factor', value: '--' },
                { label: 'ETS Price', value: '--' },
                { label: 'Year', value: 2025 },
            ];
            $('#compact-stat-row').empty();
            compact_stats.forEach(stat => {
                $('#compact-stat-row').append(`<div class='col mb-2'>${render_compact_card(stat.label, stat.value)}</div>`);
            });
        }

        /**
         * Load and render the report table and update chart.
         * @param {boolean} reset If true, reset table and start from first page.
         * @param {Object} [filters={}]
         */
        function load_report_table(reset = false, filters_arg = {}) {
            if (reset) {
                start = 0;
                all_data = [];
                if (datatable) {
                    datatable.destroy();
                    datatable = null;
                }
            }
            const selected_filters = cbam.get_all_selected_filters(filters);
            frappe.call({
                method: 'cbam.cbam.page.financial_evaluation_report.financial_evaluation_report.get_report_data',
                args: { filters: filters_arg, selected_filters, start, page_length },
                callback: function (r) {
                    if (r.message) {
                        const { columns, data, chart_data, total_count: count } = r.message;
                        total_count = count || 0;
                        if (reset) {
                            all_data = data || [];
                        } else {
                            all_data = all_data.concat(data || []);
                        }
                        const formattedColumns = columns.map(col => ({
                            id: col.fieldname,
                            name: col.label,
                            width: col.width || 150,
                            resizable: true,
                            editable: false
                        }));
                        const per_tonne = $('#table-toggle').is(':checked');
                        const table_data = get_table_data(all_data, per_tonne);
                        if (!datatable) {
                            $('#table-section').empty();
                            datatable = new DataTable('#table-section', {
                                columns: formattedColumns,
                                data: table_data,
                                layout: 'fixed',
                                stickyHeader: true,
                                inlineFilters: true,
                                scrollY: '500px',
                                scrollX: true,
                                className: 'frappe-datatable',
                                checkboxColumn: true, 
                            });
                            bind_datatable_selection_events();
                        } else {
                            datatable.refresh(table_data);
                        }
                        // Update chart (respect 'Show Selected rows' toggle)
                        const showSelected = $('#selected-rows-toggle').is(':checked');
                        let chart_data_final;
                        if (showSelected) {
                            const selectedData = get_selected_rows_data();
                            chart_data_final = cbam.get_chart_data(selectedData, per_tonne);
                        } else {
                            chart_data_final = cbam.get_chart_data(all_data, per_tonne);
                        }
                        cbam.update_chart(chart_data_final);
                        render_pagination_controls();
                    }
                },
                error: function () {
                    $('#table-section').empty().html('<div class="text-center text-muted p-4">Failed to load data</div>');
                    cbam.update_chart(null);
                    render_pagination_controls();
                }
            });
        }

        /**
         * Render pagination controls below the table.
         */
        function render_pagination_controls() {
            $('#pagination-controls').remove();
            const showing_from = total_count === 0 ? 0 : 1;
            const showing_to = all_data.length;
            const html = `
                <div id="pagination-controls" class="cbam-pagination-footer d-flex flex-wrap align-items-center justify-content-between mt-2">
                    <div class="d-flex align-items-center gap-2 flex-wrap">
                        <span class="mx-2">${__("Showing")} <b>${showing_from}&ndash;${showing_to}</b> ${__("of")} <b>${total_count}</b></span>
                        <span class="mx-2">| ${__("Rows per load")}: </span>
                        <select id="page-length-select" class="form-select form-select-sm d-inline-block w-auto">
                            <option value="20">20</option>
                            <option value="50">50</option>
                            <option value="100">100</option>
                            <option value="500">500</option>
                        </select>
                    </div>
                    <div class="ms-auto">
                        <button id="load-more" class="btn btn-primary btn-sm px-2" ${(all_data.length >= total_count) ? 'disabled' : ''}>${__("Load More")}</button>
                    </div>
                </div>
                <style>
                .cbam-pagination-footer {
                    border: 1px solid #e3e6eb;
                    background: #f8f9fa;
                    border-radius: 0.5rem;
                    padding: 0.75rem 1rem;
                    box-shadow: 0 1px 2px rgba(0,0,0,0.03);
                }
                @media (max-width: 600px) {
                    .cbam-pagination-footer { flex-direction: column; align-items: stretch; }
                    .cbam-pagination-footer .ms-auto { margin-left: 0 !important; margin-top: 0.5rem; }
                }
                </style>
            `;
            $('#table-section').after(html);
            $('#page-length-select').val(page_length);
            $('#page-length-select').on('change', function () {
                page_length = parseInt($(this).val(), 10);
                load_report_table(true); // reset and reload
            });
            $('#load-more').on('click', () => {
                start = all_data.length;
                load_report_table(false);
            });
        }

        /**
         * Render a compact stat card.
         * @param {string} label
         * @param {string|number} value
         * @returns {string}
         */
        function render_compact_card(label, value) {
            return `
                <div class="frappe-card d-flex flex-column justify-content-center align-items-center border rounded shadow-sm p-3 text-center stat-card h-100">
                    <div class="text-muted small">${label}</div>
                    <div class="fw-bold fs-5 mt-1 stat-value" data-stat="${label.toLowerCase().replace(/ /g, '-')}">${value}</div>
                </div>
            `;
        }

        /**
         * Setup table toggle functionality.
         */
        function setup_table_toggle() {
            // Table toggle functionality
            $('#table-toggle').on('change', function() {
                const isVisible = $(this).is(':checked');
                
                // You can add your custom actions here based on the toggle state
                console.log('Table toggle changed:', isVisible);
                
                // Example: Show/hide table
              
            });
        }
    })();
};
